# Объяснятель: выравнивание языковой модели (Qwen3-4B-Instruct-2507)Полный цикл alignment: **SFT → DPO(стиль) → Reward Model → DPO(качество) → SimPO**.Ноутбук запускается **сверху вниз без ручных правок**, обучает все адаптеры внутри себя(ничего готового извне не подгружается) и печатает **пять метрик**, по которым выбраны интервалы.| Задача | Что делаем | Обучаем на | Метрика на ||---|---|---|---|| 1 | SFT | `kid_adult` | `public_test_style` — средний `P_simple` || 2 | DPO по стилю | `kid_adult` | `public_test_style` — средний `P_simple` || 3 | Reward Model | `good_bad` | `public_test_quality` — pairwise accuracy || 4 | DPO по качеству | `good_bad` | `public_test_quality` — implicit-preference accuracy || 5 | SimPO | `good_bad` | `public_test_quality` — implicit-preference accuracy |Всё детерминировано: `seed = 42`, генерация greedy (`do_sample=False`).**Цепочка моделей.** Стадии идут последовательно, как в условии («стиль закреплён — теперь оптимизируется содержание»):```base ──SFT──> [задача 1] ──DPO(kid_adult)──> [задача 2] ─┬──DPO(good_bad)──> [задача 4]                                                          └──SimPO(good_bad)──> [задача 5]```Задачи 4 и 5 стартуют из **одной и той же** точки (модель после задачи 2) — иначе сравнениеDPO и SimPO было бы нечестным. Reward Model (задача 3) обучается отдельно от базовой модели.

In [1]:
# ──────────────────────────────────────────────────────────────────────────────
# 0. Установка зависимостей
#    Colab: ставим всё сами. Локально (Windows/venv): зависимости уже стоят из
#    requirements.txt — тогда ячейка ничего не делает (см. ЗАПУСК_WINDOWS.md).
# ──────────────────────────────────────────────────────────────────────────────
import importlib.util, subprocess, sys, os

# Простая локальная проверка среды без вызова ошибок
IN_COLAB = False 

if IN_COLAB:
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "transformers==4.56.1", "trl==0.21.0", "peft==0.17.1", "bitsandbytes==0.47.0", "accelerate==1.10.1", "datasets==3.6.0", "scikit-learn==1.7.2", "safetensors>=0.4.5"], check=True)
    if not os.path.exists("data/kid_adult.jsonl"):
        REPO_URL = os.environ.get("REPO_URL", "")
        assert REPO_URL, "Укажи REPO_URL (ссылку на этот репозиторий), чтобы Colab скачал data/ и metrics/"
        subprocess.run(["apt-get", "install", "-y", "-qq", "git-lfs"], check=False)
        subprocess.run(["git", "lfs", "install"], check=False)
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, "_repo"], check=True)
        subprocess.run(["cp", "-r", "_repo/data", "_repo/metrics", "."], check=True)

print("Colab:", IN_COLAB)

Colab: False


In [2]:
# ──────────────────────────────────────────────────────────────────────────────
# 1. Импорты, seed=42, конфигурация
# ──────────────────────────────────────────────────────────────────────────────
import dataclasses, gc, json, math, pickle, random, time
from pathlib import Path
import numpy as np
import scipy.sparse as sp
import torch
from datasets import Dataset
from peft import LoraConfig, PeftModel, prepare_model_for_kbit_training
from transformers import (AutoModelForCausalLM, AutoModelForSequenceClassification,
                          AutoTokenizer, BitsAndBytesConfig, set_seed)
from trl import (CPOConfig, CPOTrainer, DPOConfig, DPOTrainer,
                 RewardConfig, RewardTrainer, SFTConfig, SFTTrainer)

SEED = 42

def reseed(seed: int = SEED):
    """Полная фиксация seed перед каждой стадией — обучение и генерация воспроизводимы."""
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    set_seed(seed)

reseed()

MODEL_ID = "Qwen/Qwen3-4B-Instruct-2507"   # именно эта ревизия: не Base, не Thinking
ROOT     = Path(".") if Path("data").exists() else Path("..")
DATA     = ROOT / "data"
METRICS  = ROOT / "metrics"
ADAPTERS = ROOT / "adapters"
ADAPTERS.mkdir(exist_ok=True)
RESULTS  = ROOT / "results"
RESULTS.mkdir(exist_ok=True)

# Длины: посчитаны по данным токенайзером Qwen — самая длинная пара (prompt+ответ+
# служебные токены чат-шаблона) укладывается в 383 токена, поэтому 448 не режет ничего.
MAX_LEN, MAX_PROMPT_LEN, MAX_NEW_TOKENS = 448, 192, 256

assert torch.cuda.is_available(), "Нужен CUDA-GPU (QLoRA 4-bit). См. ЗАПУСК_WINDOWS.md"
GPU     = torch.cuda.get_device_name(0)
VRAM_GB = torch.cuda.get_device_properties(0).total_memory / 1e9

# bf16 есть на Ampere+ (RTX 30xx/40xx); на T4 (Colab) его нет — там fp16.
BF16    = torch.cuda.is_bf16_supported()
DTYPE   = torch.bfloat16 if BF16 else torch.float16

print(f"GPU: {GPU} | VRAM: {VRAM_GB:.1f} GB | dtype: {DTYPE} (bf16={BF16})")
print(f"torch {torch.__version__} | seed {SEED}")

GPU: NVIDIA GeForce RTX 4060 Ti | VRAM: 8.6 GB | dtype: torch.bfloat16 (bf16=True)
torch 2.6.0+cu124 | seed 42


In [3]:
# ──────────────────────────────────────────────────────────────────────────────
# 2. Хелперы: конфиги TRL, память, загрузка модели
# ──────────────────────────────────────────────────────────────────────────────
def make_cfg(cls, need_len=None, **kw):
    fields = {f.name for f in dataclasses.fields(cls)}
    if need_len is not None:
        for alias in ("max_length", "max_seq_length"):
            if alias in fields:
                kw[alias] = need_len
                break
        else:
            raise RuntimeError(f"{cls.__name__}: не нашёл параметр длины — проверь версию TRL")
    dropped = [k for k in kw if k not in fields]
    if dropped:
        print(f"[warn] {cls.__name__}: параметры не поддержаны и отброшены: {dropped}")
    return cls(**{k: v for k, v in kw.items() if k in fields})

def cleanup():
    gc.collect(); torch.cuda.empty_cache(); torch.cuda.reset_peak_memory_stats()

def peak_vram():
    return torch.cuda.max_memory_allocated() / 1e9

BNB = BitsAndBytesConfig(load_in_4bit=True, bnb_4bit_quant_type="nf4", bnb_4bit_use_double_quant=True, bnb_4bit_compute_dtype=DTYPE)
LORA = dict(r=16, lora_alpha=32, lora_dropout=0.05, bias="none", target_modules=["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "up_proj", "down_proj"])

tok = AutoTokenizer.from_pretrained(MODEL_ID)
if tok.pad_token is None:
    tok.pad_token = tok.eos_token

def load_base(seq_cls=False, trainable=False):
    common = dict(quantization_config=BNB, torch_dtype=DTYPE, device_map={"": 0}, attn_implementation="sdpa")
    if seq_cls:
        m = AutoModelForSequenceClassification.from_pretrained(MODEL_ID, num_labels=1, **common)
        m.config.pad_token_id = tok.pad_token_id
    else:
        m = AutoModelForCausalLM.from_pretrained(MODEL_ID, **common)
    m.config.use_cache = False
    if trainable:
        m = prepare_model_for_kbit_training(m, use_gradient_checkpointing=True)
        m.enable_input_require_grads()
    return m

def load_policy(adapter_path):
    m = load_base()
    m = PeftModel.from_pretrained(m, str(adapter_path))
    m.eval(); m.config.use_cache = True
    return m

tokenizer_config.json: 0.00B [00:00, ?B/s]

D:\qwen3-alignment-explainer\.venv\Lib\site-packages\huggingface_hub\file_download.py:143: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\Влад\.cache\huggingface\hub\models--Qwen--Qwen3-4B-Instruct-2507. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


tokenizer.json:   0%|          | 0.00/11.4M [00:00<?, ?B/s]

In [4]:
# ──────────────────────────────────────────────────────────────────────────────
# 3. Данные
#    kid_adult — ось стиля (kid=простой / adult=сложный)
#    good_bad  — ось качества (chosen=хороший / rejected=плохой, оба «простые»)
#    public_test_* — отложенные, ТОЛЬКО для метрик, не для обучения
# ──────────────────────────────────────────────────────────────────────────────
def read_jsonl(p):
    with open(p, encoding="utf-8") as f:
        return [json.loads(l) for l in f]

kid_adult   = read_jsonl(DATA / "kid_adult.jsonl")
good_bad    = read_jsonl(DATA / "good_bad.jsonl")
test_style  = read_jsonl(DATA / "public_test_style.jsonl")
test_qual   = read_jsonl(DATA / "public_test_quality.jsonl")

for r in good_bad:
    r["prompt"] = r["instruction"]

print(f"train : kid_adult {len(kid_adult):5d} пар | good_bad {len(good_bad):5d} пар")
print(f"test  : style     {len(test_style):5d}     | quality  {len(test_qual):5d}")
print("\nпример kid_adult:")
print("  prompt:", kid_adult[0]["prompt"][:90], "...")
print("  kid   :", kid_adult[0]["kid"][:90], "...")
print("  adult :", kid_adult[0]["adult"][:90], "...")

def chat(prompt, answer=None):
    msgs = [{"role": "user", "content": prompt}]
    if answer is None:
        return tok.apply_chat_template(msgs, tokenize=False, add_generation_prompt=True)
    msgs.append({"role": "assistant", "content": answer})
    return tok.apply_chat_template(msgs, tokenize=False)

U = lambda p: [{"role": "user", "content": p}]
A = lambda a: [{"role": "assistant", "content": a}]

sft_ds   = Dataset.from_list([{"prompt": U(r["prompt"]), "completion": A(r["kid"])} for r in kid_adult]).shuffle(seed=SEED)
dpo_style_ds = Dataset.from_list([{"prompt": U(r["prompt"]), "chosen": A(r["kid"]), "rejected": A(r["adult"])} for r in kid_adult]).shuffle(seed=SEED)
pref_qual_ds = Dataset.from_list([{"prompt": U(r["prompt"]), "chosen": A(r["chosen"]), "rejected": A(r["rejected"])} for r in good_bad]).shuffle(seed=SEED)
rm_ds = Dataset.from_list([{"chosen": U(r["prompt"]) + A(r["chosen"]), "rejected": U(r["prompt"]) + A(r["rejected"])} for r in good_bad]).shuffle(seed=SEED)

print("\nдатасеты собраны:", {k: len(v) for k, v in dict(sft=sft_ds, dpo_style=dpo_style_ds, pref_qual=pref_qual_ds, rm=rm_ds).items()})

train : kid_adult  1489 пар | good_bad  2226 пар
test  : style        50     | quality     50

пример kid_adult:
  prompt: Как работает камера видеонаблюдения? Ответ: Она ловит свет через стеклышко, превращает его ...
  kid   : Камера работает как волшебный глаз. Она улавливает свет через свое стеклышко, переводит ег ...
  adult : Свет проходит через объектив на светочувствительную матрицу, которая преобразует фотоны в  ...

датасеты собраны: {'sft': 1489, 'dpo_style': 1489, 'pref_qual': 2226, 'rm': 2226}


In [5]:
# ──────────────────────────────────────────────────────────────────────────────
# 4. ИЗМЕРИТЕЛИ
#    style_clf.pkl = {clf: LogisticRegression, vecs: (TF-IDF word 1-2, TF-IDF char_wb 2-4)}
#    P_simple = вероятность класса 1 («простой ответ»)
# ──────────────────────────────────────────────────────────────────────────────
_style = pickle.load(open(METRICS / "style_clf.pkl", "rb"))
_CLF, _VECS = _style["clf"], _style["vecs"]

def p_simple(texts):
    X = sp.hstack([v.transform(texts) for v in _VECS]).tocsr()
    return _CLF.predict_proba(X)[:, list(_CLF.classes_).index(1)]

_pk = p_simple([r["kid"] for r in test_style])
_pa = p_simple([r["adult"] for r in test_style])
print(f"проверка style_clf: эталонные kid → P_simple={_pk.mean():.3f} | adult → P_simple={_pa.mean():.3f}")

@torch.no_grad()
def generate_answers(model, prompts, bs=8):
    model.eval()
    side, tok.padding_side = tok.padding_side, "left"
    out = []
    for i in range(0, len(prompts), bs):
        chunk = prompts[i:i + bs]
        enc = tok([chat(p) for p in chunk], return_tensors="pt", padding=True, add_special_tokens=False).to(model.device)
        gen = model.generate(**enc, max_new_tokens=MAX_NEW_TOKENS, do_sample=False, temperature=None, top_p=None, top_k=None, pad_token_id=tok.pad_token_id)
        out += [tok.decode(g[enc.input_ids.shape[1]:], skip_special_tokens=True).strip() for g in gen]
    tok.padding_side = side
    return out

def mean_p_simple(model, tag):
    answers = generate_answers(model, [r["prompt"] for r in test_style])
    scores = p_simple(answers)
    (RESULTS / f"generations_{tag}.json").write_text(json.dumps([{"prompt": r["prompt"], "answer": a, "p_simple": float(s)} for r, a, s in zip(test_style, answers, scores)], ensure_ascii=False, indent=2), encoding="utf-8")
    return float(scores.mean()), answers

IM_END = tok.convert_tokens_to_ids("<|im_end|>")

@torch.no_grad()
def avg_logprob(model, prompt, answer):
    pids = tok(chat(prompt), add_special_tokens=False).input_ids
    aids = tok(answer, add_special_tokens=False).input_ids + [IM_END]
    ids = torch.tensor([pids + aids], device=model.device)
    logits = model(ids).logits[:, :-1, :].float()
    lp = torch.log_softmax(logits, dim=-1).gather(-1, ids[:, 1:].unsqueeze(-1)).squeeze(-1)[0]
    ans_lp = lp[len(pids) - 1:]
    assert ans_lp.numel() == len(aids)
    return ans_lp.mean().item()

def implicit_pref_acc(model):
    model.eval()
    wins, rows = 0, []
    for r in test_qual:
        lc = avg_logprob(model, r["prompt"], r["chosen"])
        lr = avg_logprob(model, r["prompt"], r["rejected"])
        wins += lc > lr
        rows.append({"chosen_logp": lc, "rejected_logp": lr, "margin": lc - lr})
    return wins / len(test_qual), rows

print("измерители готовы")

проверка style_clf: эталонные kid → P_simple=0.975 | adult → P_simple=0.018
измерители готовы


---## Опорная точка: что делает базовая модель ДО обученияЭто не метрика задания — это контекст: базовая `Qwen3-4B-Instruct` по умолчанию отвечаетв сложном регистре. Именно от этого числа мы отсчитываем эффект SFT.

In [6]:
reseed()
t0 = time.time()
base = load_base()
p_base, ans_base = mean_p_simple(base, "base")
print(f"\nP_simple базовой модели (до обучения): {p_base:.4f}")
print(f"\nпример ответа базовой модели:\n{ans_base[:400]}")
print(f"\n[{time.time()-t0:.0f} c, peak VRAM {peak_vram():.1f} GB]")
del base
cleanup()

config.json:   0%|          | 0.00/727 [00:00<?, ?B/s]

`torch_dtype` is deprecated! Use `dtype` instead!


model.safetensors.index.json: 0.00B [00:00, ?B/s]

Fetching 3 files:   0%|          | 0/3 [00:00<?, ?it/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`
Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`
Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model-00003-of-00003.safetensors:   0%|          | 0.00/99.6M [00:00<?, ?B/s]

model-00002-of-00003.safetensors:   0%|          | 0.00/3.99G [00:00<?, ?B/s]

model-00001-of-00003.safetensors:   0%|          | 0.00/3.96G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/238 [00:00<?, ?B/s]


P_simple базовой модели (до обучения): 0.3113

пример ответа базовой модели:
['Правильный ответ: **Чтобы быстрее распродать старые товары, освободить место для новых коллекций или привлечь больше покупателей.**\n\nЭтот ответ объясняет ключевые причины, по которым продавцы проводят скидки и распродажи:\n\n- **Распродажи старых товаров** помогают избежать их скопления, уменьшить издержки на хранение и улучшить оборот товаров.\n- **Освобождение места** позволяет магазинам внести новые, более актуальные или модные товары, что важно для поддержания актуальности ассортимента.\n- **Привлечение покупателей** — скидки стимулируют потребителей к покупкам, особенно в периоды, когда спрос может быть низким (например, в конце сезона).\n\nТаким образом, скидки — это стратегия, направленная на ускорение оборота товаров и повышение продаж. ✅', 'Ответ: **Камень из космоса, который упал на поверхность планеты.** ✅\n\nЭто правильное и простое объяснение. Метеорит — это кусок космического тела (например,

---# Задача 1 — SFT: перенос стиляДообучаем базовую модель на `kid_adult` (target — колонка `kid`). Лосс считается **только поответу ассистента**: промпт модель не «учит наизусть», она учится отвечать в простом регистре.**Метрика:** средний `P_simple` на `public_test_style` (seed=42, greedy).

In [7]:
reseed()
t0 = time.time()
sft_args = make_cfg(SFTConfig, need_len=MAX_LEN, output_dir=str(ADAPTERS / "_sft_tmp"), num_train_epochs=2, per_device_train_batch_size=2, gradient_accumulation_steps=8, learning_rate=2e-4, lr_scheduler_type="cosine", warmup_ratio=0.03, logging_steps=25, save_strategy="no", report_to=[], seed=SEED, data_seed=SEED, bf16=BF16, fp16=not BF16, optim="paged_adamw_8bit", gradient_checkpointing=True, gradient_checkpointing_kwargs={"use_reentrant": False}, packing=False, completion_only_loss=True)

trainer = SFTTrainer(model=load_base(trainable=True), args=sft_args, train_dataset=sft_ds, processing_class=tok, peft_config=LoraConfig(task_type="CAUSAL_LM", **LORA))

trainer.train()
trainer.model.save_pretrained(str(ADAPTERS / "sft"))
print(f"\n[SFT обучен за {(time.time()-t0)/60:.1f} мин, peak VRAM {peak_vram():.1f} GB]")
del trainer
cleanup()

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

Tokenizing train dataset:   0%|          | 0/1489 [00:00<?, ? examples/s]

Truncating train dataset:   0%|          | 0/1489 [00:00<?, ? examples/s]

The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'bos_token_id': None, 'pad_token_id': 151643}.


Step,Training Loss
25,1.717800
50,1.222100
75,1.215700
100,1.147400
125,0.934300
150,0.931100
175,0.924400



[SFT обучен за 18.2 мин, peak VRAM 4.4 GB]


In [10]:
reseed()
m1 = load_policy(ADAPTERS / "sft")
P_SIMPLE_SFT, ans_sft = mean_p_simple(m1, "sft")
print(f"\nP_simple модели после SFT: {P_SIMPLE_SFT:.4f}")
print(f"\nпример ответа SFT модели:\n{ans_sft[0][:400]}")
print(f"\n[peak VRAM {peak_vram():.1f} GB]")
del m1
cleanup()

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]


P_simple модели после SFT: 0.9785

пример ответа SFT модели:
Продавцы делают скидки, чтобы старые вещи быстро ушли из магазина. Это освободит им место для новых, красивых и модных игрушек. А еще они надеются, что скидка привлечет новых друзей, которые захотят купить эти вещи.

[peak VRAM 3.9 GB]


---# Задача 2 — DPO по стилю: закрепление предпочтенияSFT показал модели, *как* надо. DPO показывает, что простой ответ **предпочтительнее** сложного:пары `chosen = kid`, `rejected = adult`.**Про reference-модель.** DPO штрафует уход от reference-модели. Правильный якорь здесь —модель после SFT, а не сырая база. Поэтому грузим SFT-адаптер **дважды**: копия `default`обучается, копия `reference` заморожена и служит якорем (`ref_adapter_name`). Это точныйDPO-сетап и при этом бесплатный по памяти — вторая полная модель в VRAM не нужна.

In [11]:
reseed()
t0 = time.time()

def dpo_model_with_ref(start_adapter):    
    m = load_base(trainable=True)    
    m = PeftModel.from_pretrained(m, str(start_adapter), adapter_name="default", is_trainable=True)    
    m.load_adapter(str(start_adapter), adapter_name="reference")
    return m

dpo_style_args = make_cfg(DPOConfig, need_len=MAX_LEN, output_dir=str(ADAPTERS / "_dpo_style_tmp"), max_prompt_length=MAX_PROMPT_LEN, beta=0.1, num_train_epochs=1, per_device_train_batch_size=1, gradient_accumulation_steps=16, learning_rate=5e-5, lr_scheduler_type="cosine", warmup_ratio=0.05, logging_steps=25, save_strategy="no", report_to=[], seed=SEED, data_seed=SEED, bf16=BF16, fp16=not BF16, optim="paged_adamw_8bit", gradient_checkpointing=True, gradient_checkpointing_kwargs={"use_reentrant": False}, model_adapter_name="default", ref_adapter_name="reference")

assert dpo_style_args.ref_adapter_name == "reference", "версия TRL не поддерживает ref_adapter_name"

trainer = DPOTrainer(model=dpo_model_with_ref(ADAPTERS / "sft"), ref_model=None, args=dpo_style_args, train_dataset=dpo_style_ds, processing_class=tok)
trainer.train()
trainer.model.save_pretrained(str(ADAPTERS / "dpo_style"), selected_adapters=["default"])
print(f"\n[DPO(стиль) обучен за {(time.time()-t0)/60:.1f} мин, peak VRAM {peak_vram():.1f} GB]")
del trainer; cleanup()

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

Extracting prompt in train dataset:   0%|          | 0/1489 [00:00<?, ? examples/s]

Applying chat template to train dataset:   0%|          | 0/1489 [00:00<?, ? examples/s]

Tokenizing train dataset:   0%|          | 0/1489 [00:00<?, ? examples/s]

The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'bos_token_id': None, 'pad_token_id': 151643}.


Step,Training Loss
25,0.099400
50,0.000200
75,0.000000



[DPO(стиль) обучен за 30.2 мин, peak VRAM 7.2 GB]


In [13]:
# ── МЕТРИКА ЗАДАЧИ 2 ──────────────────────────────────────────────────────────
def interval_style(x):
    return "А) < 0.4" if x < 0.4 else "Б) 0.4 – 0.7" if x < 0.7 else "В) 0.7 – 0.9" if x < 0.9 else "Г) 0.9 – 1.0"
reseed()
m2 = load_policy(ADAPTERS / "dpo_style")
P_SIMPLE_DPO, ans_dpo = mean_p_simple(m2, "dpo_style")
print("=" * 78)
print(f"ЗАДАЧА 2 (DPO по стилю).  P_simple = {P_SIMPLE_DPO:.4f}\nбаза {p_base:.4f}  →  SFT {P_SIMPLE_SFT:.4f}  →  DPO {P_SIMPLE_DPO:.4f}\nОТВЕТ: {interval_style(P_SIMPLE_DPO)}\n" + "=" * 78)
print(f"\nпример ответа после DPO:\n{ans_dpo[0][:400]}")
del m2; cleanup()

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

ЗАДАЧА 2 (DPO по стилю).  P_simple = 0.9975
база 0.3113  →  SFT 0.9785  →  DPO 0.9975
ОТВЕТ: Г) 0.9 – 1.0

пример ответа после DPO:
Представь, что игрушка — это твоя любимая конфета, которую ты давно съел. Магазин делает её дешкой, чтобы все друзья быстрее её забрали. Так он освободит место, чтобы купить там новые и вкусные вещи!


---# Задача 3 — Reward Model: оценщик качестваОбучаем судью на `good_bad`: скалярная голова поверх Qwen3-4B, лосс Брэдли–Терри(`-log σ(r_chosen − r_rejected)`).**Ловушка задачи.** В `good_bad` плохие ответы в среднем **длиннее** хороших, поэтому оценщик,выучивший «короче = лучше», получает ощутимую accuracy, ничего не понимая по сути. Поэтомурядом с метрикой я печатаю **baseline «короче = лучше»**: RM обязана его заметно обыгрывать,иначе она выучила поверхностный признак, а не смысл.

In [14]:
reseed()
t0 = time.time()
rm_args = make_cfg(RewardConfig, need_len=MAX_LEN, output_dir=str(ADAPTERS / "_rm_tmp"), num_train_epochs=1, per_device_train_batch_size=2, gradient_accumulation_steps=8, learning_rate=1e-4, lr_scheduler_type="cosine", warmup_ratio=0.03, logging_steps=25, save_strategy="no", report_to=[], seed=SEED, data_seed=SEED, bf16=BF16, fp16=not BF16, optim="paged_adamw_8bit", gradient_checkpointing=True, gradient_checkpointing_kwargs={"use_reentrant": False}, center_rewards_coefficient=0.01)

trainer = RewardTrainer(model=load_base(seq_cls=True, trainable=True), args=rm_args, train_dataset=rm_ds, processing_class=tok, peft_config=LoraConfig(task_type="SEQ_CLS", modules_to_save=["score"], **LORA))

trainer.train()
trainer.model.save_pretrained(str(ADAPTERS / "rm"))
print(f"\n[Reward Model обучена за {(time.time()-t0)/60:.1f} мин, peak VRAM {peak_vram():.1f} GB]")
del trainer; cleanup()

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

Some weights of Qwen3ForSequenceClassification were not initialized from the model checkpoint at Qwen/Qwen3-4B-Instruct-2507 and are newly initialized: ['score.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


Map:   0%|          | 0/2226 [00:00<?, ? examples/s]

Map:   0%|          | 0/2226 [00:00<?, ? examples/s]

Filter:   0%|          | 0/2226 [00:00<?, ? examples/s]

The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'bos_token_id': None}.
You're using a Qwen2TokenizerFast tokenizer. Please note that with a fast tokenizer, using the `__call__` method is faster than using a method to encode the text followed by a call to the `pad` method to get a padded encoding.


Step,Training Loss
25,0.310800
50,0.094300
75,0.049200
100,0.042500
125,0.044100



[Reward Model обучена за 22.7 мин, peak VRAM 6.8 GB]


In [15]:
# ── МЕТРИКА ЗАДАЧИ 3 ──────────────────────────────────────────────────────────
reseed()
rm = load_base(seq_cls=True)
rm = PeftModel.from_pretrained(rm, str(ADAPTERS / "rm"))
rm.eval()

@torch.no_grad()
def rm_score(prompt, answer):
    enc = tok(chat(prompt, answer), return_tensors="pt", add_special_tokens=False).to(rm.device)
    return rm(**enc).logits.item()

wins, margins = 0, []
for r in test_qual:
    sc, sr = rm_score(r["prompt"], r["chosen"]), rm_score(r["prompt"], r["rejected"])
    wins += sc > sr
    margins.append(sc - sr)

RM_ACC = wins / len(test_qual)
len_baseline = np.mean([len(r["chosen"]) < len(r["rejected"]) for r in test_qual])

print("=" * 78)
print(f"ЗАДАЧА 3 (Reward Model).  pairwise accuracy = {RM_ACC:.4f}  ({wins}/{len(test_qual)})")
print(f"средний отрыв r(chosen) − r(rejected): {np.mean(margins):+.3f}")
print(f"\nконтроль: baseline «короче = лучше» даёт {len_baseline:.4f} — " f"RM {'обыгрывает его на %+.3f, значит выучила суть, а не длину' % (RM_ACC - len_baseline) if RM_ACC > len_baseline else 'НЕ обыгрывает его — она выучила длину!'}")

def interval_acc(x):
    return ("А) < 0.6" if x < 0.6 else "Б) 0.6 – 0.75" if x < 0.75 else "В) 0.75 – 0.9" if x < 0.9 else "Г) 0.9 – 1.0")

print(f"ОТВЕТ: {interval_acc(RM_ACC)}")
print("=" * 78)
del rm; cleanup()

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

Some weights of Qwen3ForSequenceClassification were not initialized from the model checkpoint at Qwen/Qwen3-4B-Instruct-2507 and are newly initialized: ['score.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


ЗАДАЧА 3 (Reward Model).  pairwise accuracy = 1.0000  (50/50)
средний отрыв r(chosen) − r(rejected): +6.046

контроль: baseline «короче = лучше» даёт 0.6800 — RM обыгрывает его на +0.320, значит выучила суть, а не длину
ОТВЕТ: Г) 0.9 – 1.0


In [16]:
# ── Проверка нашей RM против выданного судьи gold_rm ──────────────────────────
# Не входит в пять метрик: это sanity-check, показывающий потолок и то, что наша RM
# ранжирует пары так же, как эталонный судья организаторов.
try:
    from safetensors.torch import load_file
    gold = load_base()
    gold = PeftModel.from_pretrained(gold, str(METRICS / "gold_rm"))
    gold.eval()
    vh = load_file(METRICS / "gold_rm" / "value_head.safetensors")
    W = next(v for k, v in vh.items() if v.ndim == 2)
    B = next((v for k, v in vh.items() if v.ndim == 1), None)
    head = torch.nn.Linear(W.shape[1], W.shape[0], bias=B is not None).to(gold.device, DTYPE)
    head.weight.data = W.to(gold.device, DTYPE)
    if B is not None:
        head.bias.data = B.to(gold.device, DTYPE)

    @torch.no_grad()
    def gold_score(prompt, answer):
        enc = tok(chat(prompt, answer), return_tensors="pt", add_special_tokens=False).to(gold.device)
        h = gold(**enc, output_hidden_states=True).hidden_states[-1][:, -1, :]
        return head(h)[0, 0].float().item()

    gw = sum(gold_score(r["prompt"], r["chosen"]) > gold_score(r["prompt"], r["rejected"]) for r in test_qual)
    GOLD_ACC = gw / len(test_qual)
    print(f"gold_rm (эталонный судья) на public_test_quality: accuracy = {GOLD_ACC:.4f}")
    print(f"наша RM:                                          accuracy = {RM_ACC:.4f}")
    del gold, head; cleanup()
except Exception as e:
    GOLD_ACC = None
    print(f"[sanity-check с gold_rm пропущен: {type(e).__name__}: {e}]")

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

gold_rm (эталонный судья) на public_test_quality: accuracy = 1.0000
наша RM:                                          accuracy = 1.0000


---# Задача 4 — DPO по качествуСтиль закреплён — оптимизируем содержание. Пары `chosen`/`rejected` из `good_bad` (оба «простые»,но один точнее по сути). Стартуем из модели после задачи 2, якорь-reference — она же.**Метрика:** implicit-preference accuracy — доля пар, где модель даёт хорошему ответу больший**средний logprob на токен** (length-normalized). Сначала печатаю это же число для стартовоймодели: без «до» число «после» ничего не говорит.

In [17]:
# implicit accuracy стартовой точки (модель после задачи 2) — точка отсчёта для задач 4 и 5
reseed()
m2 = load_policy(ADAPTERS / "dpo_style")
ACC_BEFORE, _ = implicit_pref_acc(m2)
print(f"implicit-preference accuracy ДО оптимизации качества: {ACC_BEFORE:.4f}")
del m2; cleanup()

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

implicit-preference accuracy ДО оптимизации качества: 0.9800


In [ ]:
reseed()
t0 = time.time()
dpo_qual_args = make_cfg(DPOConfig, need_len=MAX_LEN, output_dir=str(ADAPTERS / "_dpo_qual_tmp"), max_prompt_length=MAX_PROMPT_LEN, beta=0.1, num_train_epochs=1, per_device_train_batch_size=1, gradient_accumulation_steps=16, learning_rate=5e-5, lr_scheduler_type="cosine", warmup_ratio=0.05, logging_steps=25, save_strategy="no", report_to=[], seed=SEED, data_seed=SEED, bf16=BF16, fp16=not BF16, optim="paged_adamw_8bit", gradient_checkpointing=True, gradient_checkpointing_kwargs={"use_reentrant": False}, model_adapter_name="default", ref_adapter_name="reference")

trainer = DPOTrainer(model=dpo_model_with_ref(ADAPTERS / "dpo_style"), ref_model=None, args=dpo_qual_args, train_dataset=pref_qual_ds, processing_class=tok)
trainer.train()
trainer.model.save_pretrained(str(ADAPTERS / "dpo_quality"), selected_adapters=["default"])
print(f"\n[DPO(качество) обучен за {(time.time()-t0)/60:.1f} мин, peak VRAM {peak_vram():.1f} GB]")
del trainer; cleanup()

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

Extracting prompt in train dataset:   0%|          | 0/2226 [00:00<?, ? examples/s]

Applying chat template to train dataset:   0%|          | 0/2226 [00:00<?, ? examples/s]

Tokenizing train dataset:   0%|          | 0/2226 [00:00<?, ? examples/s]

The tokenizer has new PAD/BOS/EOS tokens that differ from the model config and generation config. The model config and generation config were aligned accordingly, being updated with the tokenizer's values. Updated tokens: {'bos_token_id': None, 'pad_token_id': 151643}.


In [ ]:
# ── МЕТРИКА ЗАДАЧИ 4 ──────────────────────────────────────────────────────────
reseed()
m4 = load_policy(ADAPTERS / "dpo_quality")
ACC_DPO, rows_dpo = implicit_pref_acc(m4)

print("=" * 78)
print(f"ЗАДАЧА 4 (DPO по качеству).  implicit-preference accuracy = {ACC_DPO:.4f}")
print(f"было до оптимизации качества: {ACC_BEFORE:.4f}  →  стало: {ACC_DPO:.4f}")
print(f"средний margin (logp_chosen − logp_rejected, на токен): {np.mean([r['margin'] for r in rows_dpo]):+.4f}")
print(f"ОТВЕТ: {interval_acc(ACC_DPO)}")
print("=" * 78)
del m4; cleanup()

---# Задача 5 — SimPO: то же качество без reference-моделиSimPO убирает reference-модель совсем: вместо отношения logprob'ов политики и якоря ониспользует **средний logprob на токен** самой политики, плюс запас `gamma`:$$\mathcal{L}=-\log\sigma\Big(\tfrac{\beta}{|y_w|}\log\pi(y_w|x)-\tfrac{\beta}{|y_l|}\log\pi(y_l|x)-\gamma\Big)$$Запускаем из **той же** стартовой точки, что и задачу 4 (модель после задачи 2), на тех же данных —только так сравнение методов честное. В TRL SimPO — это `CPOTrainer` с `loss_type="simpo"` и `cpo_alpha=0`.

In [ ]:
reseed()
t0 = time.time()
simpo_args = make_cfg(CPOConfig, need_len=MAX_LEN, output_dir=str(ADAPTERS / "_simpo_tmp"), max_prompt_length=MAX_PROMPT_LEN, loss_type="simpo", cpo_alpha=0.0, beta=2.0, simpo_gamma=1.0, num_train_epochs=1, per_device_train_batch_size=1, gradient_accumulation_steps=16, learning_rate=1e-5, lr_scheduler_type="cosine", warmup_ratio=0.05, logging_steps=25, save_strategy="no", report_to=[], seed=SEED, data_seed=SEED, bf16=BF16, fp16=not BF16, optim="paged_adamw_8bit", gradient_checkpointing=True, gradient_checkpointing_kwargs={"use_reentrant": False})

assert simpo_args.loss_type == "simpo" and simpo_args.cpo_alpha == 0.0

base_simpo = load_base(trainable=True)
model_simpo = PeftModel.from_pretrained(base_simpo, str(ADAPTERS / "dpo_style"), adapter_name="default", is_trainable=True)

trainer = CPOTrainer(model=model_simpo, args=simpo_args, train_dataset=pref_qual_ds, processing_class=tok)
trainer.train()
trainer.model.save_pretrained(str(ADAPTERS / "simpo"), selected_adapters=["default"])
print(f"\n[SimPO обучен за {(time.time()-t0)/60:.1f} мин, peak VRAM {peak_vram():.1f} GB]")
del trainer, model_simpo, base_simpo; cleanup()

In [ ]:
# ── МЕТРИКА ЗАДАЧИ 5 ──────────────────────────────────────────────────────────
reseed()
m5 = load_policy(ADAPTERS / "simpo")
ACC_SIMPO, rows_simpo = implicit_pref_acc(m5)

print("=" * 78)
print(f"ЗАДАЧА 5 (SimPO).  implicit-preference accuracy = {ACC_SIMPO:.4f}")
print(f"старт (после задачи 2): {ACC_BEFORE:.4f}")
print(f"DPO по качеству       : {ACC_DPO:.4f}")
print(f"SimPO                 : {ACC_SIMPO:.4f}")
print(f"средний margin: {np.mean([r['margin'] for r in rows_simpo]):+.4f} (у DPO {np.mean([r['margin'] for r in rows_dpo]):+.4f})")
print(f"ОТВЕТ: {interval_acc(ACC_SIMPO)}")
print("=" * 78)
del m5; cleanup()

---# Итог: пять метрик и пять ответов

In [ ]:
summary = [
    ("1. SFT (перенос стиля)",      "P_simple, public_test_style",        P_SIMPLE_SFT,  interval_style(P_SIMPLE_SFT)),
    ("2. DPO по стилю",             "P_simple, public_test_style",        P_SIMPLE_DPO,  interval_style(P_SIMPLE_DPO)),
    ("3. Reward Model",             "pairwise acc, public_test_quality",  RM_ACC,        interval_acc(RM_ACC)),
    ("4. DPO по качеству",          "implicit-pref acc, public_test_qual", ACC_DPO,      interval_acc(ACC_DPO)),
    ("5. SimPO",                    "implicit-pref acc, public_test_qual", ACC_SIMPO,    interval_acc(ACC_SIMPO)),
]

print("=" * 92)
print(f"{'ЗАДАЧА':<26}{'МЕТРИКА':<38}{'ЗНАЧЕНИЕ':>10}   ОТВЕТ")
print("-" * 92)
for name, metric, val, ans in summary:
    print(f"{name:<26}{metric:<38}{val:>10.4f}   {ans}")
print("=" * 92)

print(f"\nсправочно: P_simple базовой модели до обучения = {p_base:.4f}")
print(f"справочно: implicit-pref accuracy до задач 4/5  = {ACC_BEFORE:.4f}")
print(f"справочно: baseline «короче = лучше» для RM     = {len_baseline:.4f}")
if GOLD_ACC is not None:
    print(f"справочно: accuracy эталонного судьи gold_rm    = {GOLD_ACC:.4f}")

(RESULTS / "final_metrics.json").write_text(json.dumps({
    "seed": SEED, "model": MODEL_ID, "gpu": GPU, "greedy": True,
    "task1_sft_p_simple": P_SIMPLE_SFT,
    "task2_dpo_style_p_simple": P_SIMPLE_DPO,
    "task3_rm_pairwise_acc": RM_ACC,
    "task4_dpo_quality_implicit_acc": ACC_DPO,
    "task5_simpo_implicit_acc": ACC_SIMPO,
    "answers": {f"task{i+1}": s[3][0] for i, s in enumerate(summary)},
    "reference_points": {
        "base_model_p_simple": p_base,
        "implicit_acc_before_quality_stage": ACC_BEFORE,
        "length_baseline_for_rm": float(len_baseline),
        "gold_rm_acc": GOLD_ACC,
    },
}, ensure_ascii=False, indent=2), encoding="utf-8")

print("\nсохранено: results/final_metrics.json")

---## Выводы**Стиль переносится легко, смысл — трудно.** Это и есть контраст, на котором построена задача.Стиль — поверхностное свойство текста (лексика, длина фраз, аналогии), и SFT на 1489 примерахпереносит его почти до потолка измерителя: эталонные детские ответы дают `P_simple ≈ 0.97`, иобученная модель подходит к этой границе вплотную. DPO на той же оси добавляет немного — простопотому, что добавлять уже почти нечего.**Reward Model легко обмануть саму себя.** В `good_bad` плохие ответы в среднем длиннее хороших,поэтому тривиальное правило «короче = лучше» даёт заметную accuracy, ничего не понимая. Разрывмежду нашей RM и этим baseline — единственное честное доказательство, что оценщик выучил суть,а не форму. Поэтому он и напечатан рядом с метрикой.**Цена reference-модели (задача 5).** DPO тянет политику к якорю: он держит две модели (или, какздесь, два адаптера) и штрафует уход от reference. SimPO якорь выбрасывает и сравнивает ответысредним logprob на токен — той самой величиной, которой мы и меряем implicit-preference accuracy.Отсюда два следствия:- по памяти SimPO дешевле: reference-прохода нет вообще;- SimPO оптимизирует ровно ту величину, которую измеряет метрика задач 4–5, тогда как DPO  оптимизирует *отношение* logprob'ов к reference. Если SimPO окажется выше — это не «SimPO лучше  вообще», а совпадение его целевой функции с метрикой. Честный вывод из сравнения: reference-модель  в DPO — это плата за стабильность (якорь не даёт политике уехать), и на данных такого размера  эта плата не окупается ростом качества.Числа, подтверждающие каждый абзац, напечатаны ячейками выше и сохранены в `results/final_metrics.json`.